src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs

```diff
@@ -102,7 +103,7 @@ public HttpProtocol(PipeFactory pipeFactory, IntPtr pHttpContext)
                 // var localAddress = GetSocketAddress(pHttpRequest->Request.Address.pLocalAddress);
                 // var remoteAddress = GetSocketAddress(pHttpRequest->Request.Address.pRemoteAddress);
 
-                RequestHeaders = new RequestHeaders(pHttpRequest);
+                RequestHeaders = new HeaderCollection(new RequestHeaders(new NativeRequestContext((HttpApiTypes.HTTP_REQUEST*)pHttpRequest)));
```

Comment:
3 allocations just to wrap headers? Looks like overkill no? Is this because of code sharing? Couldn't this be refactored?

Reply:
The NativeRequestContext will eventually be used to obtain all of the information about the request. I just wanted to get IISIntegration working before refactoring everything to use the NativeRequestContext.